src/btree/bt_sync_obsolete.c

```diff
@@ -291,3 +290,324 @@ __wt_sync_obsolete_cleanup(WT_SESSION_IMPL *session, WT_REF *parent)
+              session, &ref, __checkpoint_cleanup_page_skip, NULL, flags)) == 0 &&
+      ref != NULL) {
+        if (F_ISSET(ref, WT_REF_FLAG_INTERNAL))
+            WT_WITH_PAGE_INDEX(session, ret = __checkpoint_cleanup_obsolete_cleanup(session, ref));
+        WT_ERR(ret);
+
+        /* Check if we're quitting. */
+        if (!__checkpoint_cleanup_run_chk(session))
+            break;
+    }
+
+err:
+    /* On error, clear any left-over tree walk. */
+    WT_TRET(__wt_page_release(session, ref, flags));
+    WT_TRET(__wt_session_release_dhandle(session));
+    return (ret);
+}
+
+/*
+ * __checkpoint_cleanup_int --
+ *     Internal function to perform checkpoint cleanup of all eligible files.
+ */
+static int
+__checkpoint_cleanup_int(WT_SESSION_IMPL *session)
+{
+    WT_CURSOR *cursor;
+    WT_DECL_RET;
+    const char *uri;
+
+    WT_RET(__wt_metadata_cursor(session, &cursor));
+    while ((ret = cursor->next(cursor)) == 0) {
+        WT_ERR(cursor->get_key(cursor, &uri));
+
+        /* Ignore non-btree objects as well as the metadata file. */
+        if (!WT_BTREE_PREFIX(uri) || strcmp(uri, WT_METAFILE_URI) == 0)
+            continue;
+
+        ret = __checkpoint_cleanup_walk_btree(session, uri);
+        if (ret == ENOENT || ret == EBUSY) {
+            __wt_verbose_debug1(session, WT_VERB_CHECKPOINT_CLEANUP,
+              "%s: skipped performing checkpoint cleanup because the file %s", uri,
+              ret == ENOENT ? "does not exist" : "is busy");
+            continue;
+        }
+        WT_ERR(ret);
+
+        /* Wait for 5 seconds before proceeding with another table. */
+        __wt_cond_wait(
+          session, S2C(session)->cc_cleanup.cond, 5 * WT_MILLION, __checkpoint_cleanup_run_chk);
+
+        /* Check if we're quitting. */
+        if (!__checkpoint_cleanup_run_chk(session))
+            break;
+    }
+    WT_ERR_NOTFOUND_OK(ret, false);
+
+err:
+    WT_TRET(__wt_metadata_cursor_release(session, &cursor));
+    return (ret);
+}
+
+/*
+ * __checkpoint_cleanup --
+ *     The checkpoint cleanup thread.
+ */
+static WT_THREAD_RET
+__checkpoint_cleanup(void *arg)
+{
+    WT_CONNECTION_IMPL *conn;
+    WT_DECL_RET;
+    WT_SESSION_IMPL *session;
+    uint64_t last, now;
+    bool cv_signalled;
+
+    session = arg;
+    conn = S2C(session);
+
+    __wt_seconds(session, &last);
+    for (;;) {
+        /* Check periodically in case the signal was missed. */
+        __wt_cond_wait_signal(session, conn->cc_cleanup.cond, 5 * WT_MILLION,
+          __checkpoint_cleanup_run_chk, &cv_signalled);
+
+        /* Check if we're quitting. */
+        if (!__checkpoint_cleanup_run_chk(session))
+            break;
+
+        __wt_seconds(session, &now);
+
+        /*
+         * See if it is time to checkpoint cleanup. Checkpoint cleanup is an operation that
+         * typically has long intervals so skipping some should have little impact.
+         */
+        if (!cv_signalled && (now - last < 60))
+            continue;
```

Comment:
Should we decrease the frequency or make it configurable?

Reply:
I would prefer it to be configurable so if we have any issue with this change we have a means to adjust.

Reply:
Done.